ext/dom/xml_serializer.c

```diff
@@ -610,32 +684,54 @@ static int dom_xml_serialize_cdata_section_node(xmlOutputBufferPtr out, xmlNodeP
 	return xmlOutputBufferWriteLit(out, "]]>");
 }
 
+static zend_string *dom_xml_create_localname_set_key(const xmlAttr *attr)
+{
+	if (attr->ns == NULL || attr->ns->href == NULL) {
+		return zend_string_init((const char *) attr->name, strlen((const char *) attr->name), false);
+	}
+
+	return zend_string_concat3(
+		(const char *) attr->ns->href, strlen((const char *) attr->ns->href),
+		"", 1, /* include the \0 */
+		(const char *) attr->name, strlen((const char *) attr->name)
+	);
```

Comment:
We actually need to store tuples of (href, name) inside the hash map. However, we don't have native support for tuples in HashTable. So the trick is to concatenate the href and name, separated by \0, to use as a key instead. This achieves the same result. It only works because href and name cannot contain an embedded \0 byte. I'll add some comments to explain how this works in the code.

Reply:
Right, comments would be quite helpful here :)